rcl/test/rcl/test_events.cpp

```diff
@@ -0,0 +1,386 @@
+  rcl_node_t * node_ptr;
+
+  rcl_publisher_t publisher;
+  rcl_event_t publisher_event;
+  rcl_subscription_t subscription;
+  rcl_event_t subscription_event;
+};
+
+bool
+wait_for_msgs_and_events(
+  rcl_subscription_t * subscription,
+  rcl_event_t * subscription_event,
+  rcl_event_t * publisher_event,
+  rcl_context_t * context,
+  size_t max_tries,
+  int64_t period_ms,
+  bool & msg_ready,
+  bool & subscription_event_ready,
+  bool & publisher_event_ready)
+{
+  msg_ready = false;
+  subscription_event_ready = false;
+  publisher_event_ready = false;
+
+  int num_subscriptions = (nullptr == subscription ? 0 : 1);
+  int num_events = (nullptr == subscription_event ? 0 : 1) + (nullptr == publisher_event ? 0 : 1);
+
+  rcl_wait_set_t wait_set = rcl_get_zero_initialized_wait_set();
+  rcl_ret_t ret = rcl_wait_set_init(&wait_set, num_subscriptions, 0, 0, 0, 0, num_events,
+    context, rcl_get_default_allocator());
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
+    rcl_ret_t ret = rcl_wait_set_fini(&wait_set);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  size_t iteration = 0;
+  do {
+    ++iteration;
+    ret = rcl_wait_set_clear(&wait_set);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    if (nullptr != subscription) {
+      ret = rcl_wait_set_add_subscription(&wait_set, subscription, NULL);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    }
+    if (nullptr != subscription_event) {
+      ret = rcl_wait_set_add_event(&wait_set, subscription_event, NULL);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    }
+    if (nullptr != publisher_event) {
+      ret = rcl_wait_set_add_event(&wait_set, publisher_event, NULL);
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    }
+
+    ret = rcl_wait(&wait_set, RCL_MS_TO_NS(period_ms));
+    if (ret == RCL_RET_TIMEOUT) {
+      continue;
+    }
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+    for (size_t i = 0; i < wait_set.size_of_subscriptions; ++i) {
+      if (wait_set.subscriptions[i] && wait_set.subscriptions[i] == subscription) {
+        msg_ready = true;
+      }
+    }
+    for (size_t i = 0; i < wait_set.size_of_events; ++i) {
+      if (nullptr != wait_set.events[i]) {
+        if (wait_set.events[i] == subscription_event) {
+          subscription_event_ready = true;
+        } else if (wait_set.events[i] == publisher_event) {
+          publisher_event_ready = true;
+        }
+      }
+    }
+  } while (iteration < max_tries);
+
+  return subscription_event_ready || publisher_event_ready;
+}
+
+/*
+ * Basic test of publisher and subscriber liveliness events
+ */
+TEST_F(CLASSNAME(TestEventFixture, RMW_IMPLEMENTATION), test_pubsub_liveliness)
+{
+  SetUpPubSub(RCL_PUBLISHER_LIVELINESS_LOST, RCL_SUBSCRIPTION_LIVELINESS_CHANGED);
+
+  // TODO(wjwwood): add logic to wait for the connection to be established
```

Comment:
Why have the tools to do this properly now, let's not build in this issue from the get go. You can use `rcl_subscription_get_publisher_count` (https://github.com/ros2/rcl/blob/64a1412594b147bc363146e02d5ebc065c0d8a59/rcl/include/rcl/subscription.h#L415-L440) and/or `rcl_publisher_get_subscription_count` (https://github.com/ros2/rcl/blob/64a1412594b147bc363146e02d5ebc065c0d8a59/rcl/include/rcl/publisher.h#L441-L466) to check in a busy loop, allowing you to wait a long time if needed, but a very short time most of the time.

Something like:

```c++
// total wait time of 10 seconds, if never ready
size_t max_iterations = 100;
std::milliseconds wait_period(10);
size_t iteration = 0;
do {
  iteration++;
  size_t count = 0;
  rcl_ret_t ret = rcl_subscription_get_publisher_count(&subscription, &count);
  ASSERT_EQ(RCL_RET_OK, ret);
  if (count > 0) {
    break;
  }
  std::this_thread::sleep_for(wait_period);
} while (iteration < max_iterations);
```

Reply:
Added